pipelining_example/pipelining_example.c

```diff
@@ -0,0 +1,2793 @@
+	int gid_idx;		/* gid index to use */
+	int io_size;
+	int queue_depth;
+	enum test_type type;
+	int time; /* test time in seconds */
+	bool signature_calc;
+	enum ibv_mtu mtu;
+	const struct signature_ops *sig;
+} conf = {
+	.dev_name 	= NULL,
+	.server_name 	= NULL,
+	.tcp_port 	= 19875,
+	.ib_port 	= 1,
+	.gid_idx 	= -1,
+	.io_size 	= 16384,
+	.queue_depth 	= 1,
+	.type 		= TEST_TYPE_WRITE,
+	.time 		= 10,
+	.signature_calc = true,
+	.mtu 		= IBV_MTU_1024,
+	.sig 		= &sig_ops[SIG_TYPE_CRC32],
+};
+
+/* structure to exchange data which is needed to connect the QPs */
+struct cm_con_data_t {
+	uint16_t lid;
+	uint8_t gid[16];
+	uint32_t qp_num;
+} __attribute__((packed));
+
+enum msg_types {
+	MSG_TYPE_READ_REQ = 0,
+	MSG_TYPE_READ_REP,
+	MSG_TYPE_WRITE_REQ,
+	MSG_TYPE_WRITE_REP,
+	MSG_TYPE_STOP_REQ,
+};
+
+struct iov {
+	uint64_t addr;
+	uint32_t length;
+	uint32_t rkey;
+} __attribute__((packed));
+
+struct msg_req_hdr {
+	uint16_t type;
+	uint64_t id;
+	uint16_t iov_size;
+	uint32_t inline_data_size;
+} __attribute__((packed));
+
+enum msg_rep_status {
+	MSG_REP_STATUS_OK = 0,
+	MSG_REP_STATUS_FAIL,
+};
+
+struct msg_rep_hdr {
+	uint16_t type;
+	uint64_t id;
+	uint16_t status;
+	uint32_t inline_data_size;
+} __attribute__((packed));
+
+#define MAX_SEND_WRS 4
+#define CQ_SIZE ((MAX_SEND_WRS + 2) * conf.queue_depth)
+
+#define REPEATED_BLOCK_ELEMENTS 1
+#define RDMA_SGL_SIZE 4
+
+#define IOV_MAX_SIZE 1
+
+#define MSG_REQ_HDR_SIZE (sizeof(struct msg_req_hdr))
+#define MSG_REQ_MAX_SIZE (MSG_REQ_HDR_SIZE + MSG_MAX_INLINE_DATA_SIZE)
+
+#define MSG_REP_HDR_SIZE (sizeof(struct msg_rep_hdr))
+#define MSG_REP_MAX_SIZE (MSG_REP_HDR_SIZE + MSG_MAX_INLINE_DATA_SIZE)
+#define MSG_MAX_INLINE_DATA_SIZE 4096
+
+#define MAX_IO_SIZE (1024 * 1024)
+
+/*
+ * Data structure in server
+ * ###################################################
+ * # Data (4096)                                  #PI#
+ * ###################################################
+ * # Data (4096)                                  #PI#
+ * ###################################################
+ * # Data (4096)                                  #PI#
+ * ###################################################
+ */
+#define SERVER_DATA_SIZE 4096
+#define SERVER_MAX_BLOCKS_PER_IO ((MAX_IO_SIZE - 1) / SERVER_DATA_SIZE + 1)
+#define CLIENT_DATA_BUFFER_SIZE MAX_IO_SIZE
+
+struct buffer {
+	uint8_t *ptr;
+	size_t size;
+	size_t mem_size;
+	struct ibv_mr *mr;
+};
+
+struct ind_buffer {
+	struct buffer *orig;
+	size_t offset;
+	size_t size;
+};
+
+enum task_types {
+	TASK_TYPE_READ,
+	TASK_TYPE_WRITE
+};
+
+enum task_status {
+	TASK_STATUS_FREE = 0,
+	TASK_STATUS_INITED,
+	TASK_STATUS_RDMA_READ_DONE,
+	TASK_STATUS_RDMA_WRITE_DONE,
+	TASK_STATUS_IO_DONE,
+	TASK_STATUS_REPLY_SENT,
+	TASK_STATUS_WR_CANCELED_WITH_ERR,
+};
+
+struct repeated_block {
+	size_t prev_suffix;
+	size_t prefix;
+	size_t data_size;
+	size_t signature_size;
+	size_t repeated_count;
+	size_t stride;
+	size_t suffix;
+};
```

Comment:
This is related to one specific use-case. I think it is redundant and confusing for the example. Let's make the example as small and simple as possible. I suggest to drop this and assume that each IO is aligned to block size and offset = 0.

Reply:
Ok, I will simplify it, thanks!

Reply:
For simplicity, can I set both SERVER_DATA_SIZE and block size to be 4096? So I need only to configure the number of blocks, and the IO size will be (config.nb * SERVER_DATA_SIZE).

Reply:
That's a good idea.